### PR TITLE
Predictive feature models and directory reorg

### PR DESCRIPTION
**Pull request by bamader**
## Background
This PR does a couple of things:
1. creates and fills the predictive feature directory, which contains data files and scripts that are used for some basic data analysis devoted to identifying features that can help predict various facets of OEC care and unmet needs. This uses our new demand estimates to get a more reliable benchmark of gaps in service provision, and constructs an SVM to find demand-side characteristics that are predictive of need gaps as well as a Lasso Linear Regression to find supply-side characteristics that are associated with investment in childcare service supply (it may sound confusing here but I have awesome plans for these later).
2. adds in the stuff used for capacity above expectation calculation in the `unmet_needs` directory so that we have it and its shared across the repo
3. makes some modifications to the census data demographic fields that we retrieve to collect more data (since we needed a richer feature set for our predictive analysis), which leads to a new CSV being made there

## Validation Plan
No automated validation, but soooo much manual tweaking and testing between looking at descriptive statistics of features (Chi^2, mutual information, correlation scores, covariance, etc.) and testing different combinations of fields to train on or not. Turns out machine learning is complicated (whaaaaaaaaaat).

## Additional Context
I don't think there's any required context you need to know for reviewing this, but two observations here:
1. I opted to use a 'demand gap' instead of capacity above expectation or the eligible children metric because it made the most intuitive sense to me and because it gave the best models. Using CAE wound up bad because towns could have enough slots to satisfy their estimated demand but still grade negatively in this metric if they didn't have _enough_ of a surplus, which seemed kinda dumb. Using eligible children proved not super productive because it ended up predicting features associated with essentially number of kids, which was a little too basic to be insightful. The need gap was the place where the richest associations were.
2. This PR packs in a _lot_ of script stuff, but some of it is just retrofitting what we've already been using (the unmet needs CAE files, the census data files, etc.). The bulk of the review effort should imo be on the predictive models. And, as far as those models are concerned, I realize there are still some commented lines included in the code about dropping various features; I dropped most of the experimental crap from the code (descriptive stats and such), but left some lines in in case people wanted to change the response variables or tweak field inclusion on their own.

### CODE REVIEW

**Review by kmagida110 — APPROVED** [review]
I have a bunch of comments/questions but nothing blocking. I am curious about a few things and have comments there.

**Review by imcurtis — APPROVED** [review]

File: src/predictive_features/unmet_needs_correlation.py [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+def generate_supply_demand_dataset(supply_fp, demand_fp, supply_cols, demand_col):
+    supply_df = pd.read_csv(supply_fp)
+    demand_df = pd.read_csv(demand_fp)
+    
+    # Only need to keep City so we can join to the demographic cols
+    keep_cols = ['City']
+    keep_cols.extend(supply_cols)
+    supply_df.drop(columns=[c for c in supply_df.columns if not c in keep_cols], inplace=True)
+    supply_df.rename(columns={'City': 'town'}, inplace=True)
+    supply_df[supply_cols] = supply_df[supply_cols].apply(pd.to_numeric, errors="coerce")
```
[the remark this answers is not in the document]
Comment (bamader) [reply src/predictive_features/unmet_needs_correlation.py:52]: Not sure if some of the data was being parsed as an object, or blanks were being parsed as strings, but without the numeric coercion the script wasn't working as expected. This fixed that, most likely by the forced type coercion in the event of a failure.

File: src/predictive_features/unmet_needs_correlation.py [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+    
+    # Aggregate childcare supply between infants/toddlers and preschoolers
+    supply_df['childcare_supply'] = supply_df[supply_cols[0]] + supply_df[supply_cols[1]]
+    mean_supply = supply_df['childcare_supply'].dropna().mean()
```
[the remark this answers is not in the document]
Comment (bamader) [reply src/predictive_features/unmet_needs_correlation.py:56]: It wasn't failing but the mean was different

File: src/predictive_features/unmet_needs_correlation.py [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+estimate_df.drop(columns=[c for c in estimate_df.columns if not c in ['town', RESPONSE_VARIABLE]], inplace=True)
+
+# Read demographics file
+demos = pd.read_csv(DEMO_FILE)
+df = demos.merge(estimate_df, on='town')
+df.drop(columns=['town'], inplace=True)
+
+# Re-parse the categorical variable to be numeric
+min_house_year = df.loc[df['median year housing unit was built'] > 0, 'median year housing unit was built'].min()
+df['median year housing unit was built'] = df['median year housing unit was built'].apply(lambda x: x - min_house_year if x >= min_house_year else 0)
```
[the remark this answers is not in the document]
Comment (bamader) [reply src/predictive_features/unmet_needs_correlation.py:137]: It was easier to just rename the one column instead of adding a new column and then dropping the old one. We don't want to keep the categorical formulation of the variable around, so this is just changing the information type without losing any information content.

File: src/predictive_features/unmet_needs_correlation.py [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+# Heating fuel columns--interestingly, fuel type is actually a really
+# informative proxy of family well-being due to "purity"/"cleanliness"
+# of the fuel, moreso than cost (no fuel << wood << coal < oil, etc.) 
+# df.drop(columns=[c for c in df.columns if 'heating fuel' in str(c)], inplace=True)
+
+# Race columns--not enough training data to disambiguate the effects of different
+# racial groups; clearer results if separate race along white/non-white
+df['non-white population'] = 0
+race_cols = ['black or african american', 'american indian and alaska native', 'asian', 'native hawaiian and other pacific islander', 'other race']
+for c in race_cols:
```
[the remark this answers is not in the document]
Comment (bamader) [reply src/predictive_features/unmet_needs_correlation.py:197]: Ooooh didn't know you could aggregate columns like that in pandas

File: src/predictive_features/unmet_needs_correlation.py [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+print('Lasso correlation', reg.score(X, supply_Y))
+coef_scores = {}
+col_list = list(X.columns)
+for i in range(len(col_list)):
+    coef_scores[col_list[i]] = reg.coef_[i]
+supply_side_vars = list(coef_scores.items())
+supply_side_vars.sort(key=lambda x: abs(x[1]), reverse=True)
+supply_side_vars = [t for t in supply_side_vars if abs(t[1]) != 0]
+print('Lasso features selected', len(supply_side_vars))
+for t in supply_side_vars:
```
[the remark this answers is not in the document]
Comment (bamader) [reply src/predictive_features/unmet_needs_correlation.py:249]: Yup we do keep them around--they get written out to a file at the very end of the script. I have some plans to use these to showcase the duality of the unmet needs problem from both the supply and the demand side in a new dashboard.

File: src/predictive_features/unmet_needs_correlation.py [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+    print(t[0], "coef", t[1])
+print("")
+
+# Get a train-test split for some ML analysis
+print('Number of SVM features: ' + str(len(list(X.columns))))
+X_t, X_v, Y_t, Y_v = train_test_split(X, Y, test_size=0.3)
+X_t2, X_v2, Y_t2, Y_v2 = train_test_split(X, Y_c, test_size=0.3, stratify=Y_c)
+
+# Do a little hyper parameter search on an SVM
```
[the remark this answers is not in the document]
Comment (bamader) [reply src/predictive_features/unmet_needs_correlation.py:258]: Yeah I contemplated grid search here, but I found that cross validation really messed up all facets of my analysis here. Because of the really small data set size (< 170 examples) coupled with class imbalance, trying to do any kind of CV split led to wildly fluctuating and really numerically unstable accuracy values on each of the validations. A model with a large training set and a stratified validation might score the 'true' accuracy as 0.65, but even for 3-fold validation these accuracies would fluctuate as wildly as [0.1, 0.4, 0.8], and obviously that average is well below the larger model prediction. So I opted to not use any CV based functions (which is also why I didn't use Python's RFECV to find the optimal number of features and wrote out my own thing instead--I compared those results and the "optimal" number of features the RFECV chooser picked was always between 3 and 7 because it just threw out a bunch of features until it reached something whose average didn't bounce around everywhere).

File: src/predictive_features/unmet_needs_correlation.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (kmagida110) [reply src/predictive_features/unmet_needs_correlation.py:56]: That is really strange, mean should drop na by default? I blame the ghosts in the machine

File: src/predictive_features/unmet_needs_correlation.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (kmagida110) [reply src/predictive_features/unmet_needs_correlation.py:137]: Makes sense

File: src/predictive_features/unmet_needs_correlation.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (kmagida110) [reply src/predictive_features/unmet_needs_correlation.py:197]: Yeah, pandas can do some really cool things, and eat bamboo
